Cron/Images.php

```diff
@@ -0,0 +1,236 @@
+    public function execute()
+    {
+        // validate configuration:
+        if (!$this->config->isModuleEnabled() || !$this->config->isImagesSyncEnabled()) {
+            return;
+        }
+        // start process:
+        $this->logger->info('Synchronizing Stockbase images...');
+        // get all the eans:
+        $allEans = $this->getAllEans();
+        // get processed eans:
+        $processedEans = json_decode($this->getImagesEans()) ?: array();
+        // process 100 unprocessed eans at a time:
+        $eans = array_slice(array_diff($allEans, $processedEans), 0, 100);
+        try {
+            $client = $this->stockbaseClientFactory->create();
+            $images = $client->getImages($eans);
+        } catch (Exception $e) {
+            $this->logger->info('Cron runImageImport error: '.$e->getMessage());
+            return false;
+        }
+        // download and save the images locally:
+        $this->saveImageForProduct($images);
+        // update the processed images configuration:
+        $processedEans = array_merge($processedEans, $eans);
+        $encodedEans = json_encode($processedEans);
+        $this->saveImagesEans($encodedEans);
+        $this->logger->info('Stockbase images synchronization complete.');
+    }
+
+    /**
+     * @return array
+     */
+    public function getAllEans()
+    {
+        // get ean attribute:
+        $attribute = $this->config->getEanFieldName();
+        // apply filters and paginate by 100:
+        $collection = $this->productCollection->addAttributeToSelect($attribute)
+            ->addAttributeToFilter($attribute, array('notnull' => true, 'neq' => '')) // not null and not empty.
+            ->setPageSize(100);
+        // iterate over the pages:
+        $eans = array();
+        $currentPage = 0;
+        $lastPage = $collection->getLastPageNumber();
+        while ($currentPage < $lastPage) {
+            // load the data of this page in a single query:
+            $collection->setCurPage(++$currentPage);
+            $collection->load();
+            // iterate over the products:
+            foreach ($collection as $product) {
+                if ($ean = $product->{$attribute}) {
+                    // add the ean if this product has one:
+                    $eans[] = $ean;
+                }
+            }
+        }
+        return $eans;
+    }
+
+    /**
+     * Saves images array from stockbase for given $ean
+     *
+     * @param array $images
+     *
+     * @return bool
+     */
+    public function saveImageForProduct($images)
+    {
+        // get product model:
+        $productModel = $this->product;
+        // get ean attribute:
+        $eanField = $this->config->getEanFieldName();
+        // get client:
+        $client = $this->stockbaseClientFactory->create();
+        // loop images:
+        foreach ($images as $image) {
+            // load product by ean:
+            $product = $productModel->loadByAttribute($eanField, $image->EAN);
```

Comment:
Potential for improvement: we should be making a single query (or batch queries) to fetch many products at once for processing.

Reply:
I think this was developed like that in the M1 version since this CRON process runs every 30min and they the original intention was to not overload the server with a large batch of products, also to give time to the get image process to create the files in the Magento server.
Still it can be improved, as far we know, we can run a large process because Magento shouldn't run the same cron process again if the first process is still working.